### ISSUE TITLE
Every subexpression's type should be checked for assignability to the greatest closure of its context

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#31792]
In https://github.com/dart-lang/sdk/issues/31776#issuecomment-355711832, @leafpetersen says that the intention of the local type inference proposal is that in general, subexpressions should be checked for assignability to the greatest closure of the subexpression's context.  If the assignability criterion is met because the subexpression's type is a subtype of the greatest closure of the context, then no further action needs to be taken.  If not, then either an implicit downcast needs to be inserted, or an error should be reported.

This is slightly different from what was implemented in the front end.  The front end only checks assignability (and only inserts implicit downcasts) where necessary for soundness.  In most cases, there is no observable difference, since usually the contexts used in type inference are precisely those that are necessary for soundness.  However, there are a few exceptions, for example:

```dart
  Object o;
  int x = o..toString(); // inferred as `(o..toString()) as int`; should be (o as int)..toString()
  int y = o == x ? o : x; // inferred as (o == x ? o : x) as int; should be o == x ? o as int : x
```

**Comment by lrhn** [comment]
So, more importantly, the following would be valid:
```dart
Object o;
int x = o..toDouble();
```
Correct ?

**Comment by stereotype441** [comment]
@lrhn Yes.  Your example is currently rejected by the front end and the analyzer, and as I understand it under Leaf's proposed interpretation, it would be interpreted as `int x = (o as int)..toDouble();`, which is valid.

**Comment by leafpetersen** [comment]
Yes, that's the intention.  We actually had an issue filed about this from a user who was surprised not to get good completions from code exactly like the above, and I think it's the right thing to do.

**Comment by bwilkerson** [comment]
@danrubel It would be good for us to capture this as a test case for code completion.

**Comment by stereotype441** [comment]
It turns out that I don't have the cycles to work on this right now, so I'm unassigning.  FE team, if you want to discuss implementation strategies, please feel free to drop me a line and we can set up a VC.
